cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java

```diff
@@ -372,19 +372,33 @@ private void writeRow(Tuple2<DecoratedKey, Object[]> keyAndRowData,
     private Map<String, Object> getBindValuesForColumns(Map<String, Object> map, String[] columnNames, Object[] values)
     {
         Preconditions.checkArgument(values.length == columnNames.length,
-                                    "Number of values does not match the number of columns " + values.length + ", " + columnNames.length);
+                "Number of values does not match the number of columns " + values.length + ", " + columnNames.length);
 
         for (int i = 0; i < columnNames.length; i++)
         {
-            if (cqlTable.containsUdt(columnNames[i]))
+            String columnName = columnNames[i];
+            Object columnValue = values[i];
+
+            // convert all BridgeUDTValue types in the column to UDTValue
+            if (cqlTable.containsUdt(columnName))
             {
-                map.put(columnNames[i], maybeConvertUdt(values[i]));
+                columnValue = maybeConvertUdt(columnValue);
             }
-            else
+
+            // Find CqlTuple associated with the field and convert value to TupleValue
+            // Note: Tuples do not need recursive conversions
+            //       Tuple with UDT come here as Object[...UDTValue...]
+            //       Tuple with Tuple come here as Object[...Object[]...]
+            //       UDT with Tuple come here as UDTValue[...Object[]...]
+            CqlField.CqlTuple cqlTuple = cqlTable.findTuple(columnName);
+            if (cqlTuple != null && columnValue != null)
             {
-                map.put(columnNames[i], values[i]);
+                columnValue = cqlTuple.convertForCqlWriter(columnValue, writerContext.bridge().getVersion(), false);
             }
```

Comment:
Extracting the tuple from column value does not work. 

Taking the same `tupleList list<tuple<text, text>>` example, the column value is a list (`ArrayList`) of tuples (`Object[]`). The column value cannot be converted as if it is tuple. (It is not!)